wda/w3c_actions.py

```diff
@@ -0,0 +1,112 @@
+# coding: utf-8
+
+from typing import Any, Dict, List, Optional, Union
+
+
+class FingerMovement:
+    def __init__(self):
+        self.__data: Dict[str, Any] = {
+            "type": "pointerMove"
+        }
+    
+    @property
+    def data(self) -> Dict[str, Any]:
+        return self.__data
+    
+    def with_xy(self, x: Union[int, float], y: Union[int, float]) -> "FingerMovement":
+        self.__data["x"] = x
+        self.__data["y"] = y
+        return self
+    
+    def with_origin(self, element_uid: Optional[str]=None) -> "FingerMovement":
+        if element_uid is not None:
+            self.__data["origin"] = element_uid
+        return self
+    
+    def with_duration(self, second: Optional[float]=None) -> "FingerMovement":
+        if second is not None:
+            self.__data["duration"] = second * 1000
+        return self
+
+
+class FingerAction:
+    def __init__(self):
+        self.__data: List[Dict[str, Any]] = []
+
+    @property
+    def data(self) -> List[Dict[str, Any]]:
+        return self.__data
+    
+    def move(self, movement: FingerMovement) -> "FingerAction":
+        self.__data.append(movement.data)
+        return self
+    
+    def down(self) -> "FingerAction":
+        self.__data.append({"type": "pointerDown"})
+        return self
+    
+    def up(self) -> "FingerAction":
+        self.__data.append({"type": "pointerUp"})
+        return self
+    
+    def pause(self, second: float=0.5) -> "FingerAction":
+        self.__data.append({"type": "pause", "duration": second * 1000})
+        return self
+
+
+class W3CActions:
+    def __init__(self):
+        self.__data: List[Dict[str, Any]] = []
+    
+    @property
+    def data(self) -> List[Dict[str, Any]]:
+        return self.__data
+
+    def send_keys(self, text: str) -> "W3CActions":
+        keyboard: Dict[str, Any] = {
+            "type": "key",
+            "id": f"keyboard{len(self.__data)}",
+            "actions": [
+                (a for a in [
+                    {"type": "keyDown", "value": v},
+                    {"type": "keyUp", "value": v}
+                ]) for v in text
+            ]
+        }
+        self.__data.append(keyboard)
+        return self
+    
+    def inject_touch_actions(self, *actions: FingerAction) -> "W3CActions":
+        for action in actions:
+            pointer: Dict[str, Any] = {
+                "type": "pointer",
+                "id": f"finger{len(self.__data)}",
+                "parameters": {
+                    "pointerType": "touch"
+                },
+                "actions": action.data
+            }
+            self.__data.append(pointer)
+        return self
+    
+    def tap(self, x: Union[int, float], y: Union[int, float], element_uid: Optional[str]=None) -> "W3CActions":
+        movement = FingerMovement().with_xy(x, y).with_origin(element_uid)
+        action = FingerAction().move(movement).down().pause(0.1).up()
```

Comment:
should put 0.1 to global variable, eg: TAP_DURATION

Reply:
added `second` parameter and `0.1` as default value so as to be consistent with the `press` method